File: packages/runtime/src/server/project-dir-session.ts
export function buildDirSessionMap(
  sessions: Iterable<{ dir: string; name: string }>,
): Map<string, string[]> {
  const map = new Map<string, string[]>();

  for (const { dir, name } of sessions) {
    if (!dir) continue;
    const existing = map.get(dir);
    if (existing) {
      if (!existing.includes(name)) existing.push(name);
      continue;
    }
    map.set(dir, [name]);
  }

  return map;
}

function uniqueMatch(matches: Set<string>): string | null {
  return matches.size === 1 ? [...matches][0]! : null;
}

export function resolveSessionForProjectDir(
  projectDir: string,
  dirSessionMap: Map<string, string[]>,
): string | null {
  const exactMatches = new Set<string>(dirSessionMap.get(projectDir) ?? []);
  const exact = uniqueMatch(exactMatches);
  if (exact || exactMatches.size > 1) return exact;

  const relatedMatches = new Set<string>();
  for (const [dir, sessions] of dirSessionMap) {
    if (!projectDir.startsWith(dir + "/") && !dir.startsWith(projectDir + "/")) continue;
    for (const session of sessions) relatedMatches.add(session);
  }
  const related = uniqueMatch(relatedMatches);
  if (related || relatedMatches.size > 1) return related;

  if (!projectDir.startsWith("__encoded__:")) return null;

  const encoded = projectDir.slice("__encoded__:".length);
  const encodedMatches = new Set<string>();
  for (const [dir, sessions] of dirSessionMap) {
    if (dir.replace(/[/._]/g, "-") !== encoded) continue;
    for (const session of sessions) encodedMatches.add(session);
  }

  return uniqueMatch(encodedMatches);
}

File: packages/runtime-rs/src/project_dir_session.rs
use std::collections::{BTreeMap, BTreeSet};

pub type DirSessionMap = BTreeMap<String, Vec<String>>;

pub fn build_dir_session_map(
    sessions: impl IntoIterator<Item = (String, String)>,
) -> DirSessionMap {
    let mut map = BTreeMap::new();
    for (name, dir) in sessions {
        if dir.is_empty() {
            continue;
        }
        let names: &mut Vec<String> = map.entry(dir).or_default();
        if !names.contains(&name) {
            names.push(name);
        }
    }
    map
}

pub fn resolve_session_for_project_dir(
    project_dir: &str,
    dir_session_map: &DirSessionMap,
) -> Option<String> {
    let exact_matches = dir_session_map
        .get(project_dir)
        .map(|sessions| sessions.iter().cloned().collect::<BTreeSet<_>>())
        .unwrap_or_default();
    if !exact_matches.is_empty() {
        return unique_match(exact_matches);
    }

    let mut related_matches = BTreeSet::new();
    for (dir, sessions) in dir_session_map {
        if !project_dir.starts_with(&format!("{dir}/"))
            && !dir.starts_with(&format!("{project_dir}/"))
        {
            continue;
        }
        related_matches.extend(sessions.iter().cloned());
    }
    if !related_matches.is_empty() {
        return unique_match(related_matches);
    }

    let encoded = project_dir.strip_prefix("__encoded__:")?;

    let mut encoded_matches = BTreeSet::new();
    for (dir, sessions) in dir_session_map {
        if encode_project_dir(dir) != encoded {
            continue;
        }
        encoded_matches.extend(sessions.iter().cloned());
    }
    unique_match(encoded_matches)
}

fn unique_match(matches: BTreeSet<String>) -> Option<String> {
    if matches.len() == 1 {
        matches.into_iter().next()
    } else {
        None
    }
}

fn encode_project_dir(dir: &str) -> String {
    dir.chars()
        .map(|ch| {
            if matches!(ch, '/' | '.' | '_') {
                '-'
            } else {
                ch
            }
        })
        .collect()
}
